### PR TITLE
chapters/io: Add checker infrastructure for lab 9

### PR DESCRIPTION
**Pull request by robertpaulp**
# Prerequisite Checklist

- [x] Read the [contribution guidelines](https://github.com/cs-pub-ro/operating-systems/blob/main/CONTRIBUTING.md#pull-requests) regarding submitting new changes to the project;
- [x] Tested your changes against relevant architectures and platforms;
- [x] Updated relevant documentation (if needed).

### CODE REVIEW

**Review by Viktor09 — APPROVED** [review]
Looks good!

File: chapters/io/file-descriptors/drills/tasks/mmap_cp/checker.sh [hunk 1]
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
+# SPDX-License-Identifier: GPL-2.0
+
+# Create a test file
+echo "This is a test file." > testfile.txt
+
```
Comment (teodutu) [inline chapters/io/file-descriptors/drills/tasks/mmap_cp/checker.sh:6]:
```suggestion
Add a `make` command that compiles the source code so students only need to run one command (`./checker.sh`) to rule them all.
```

File: chapters/io/file-descriptors/drills/tasks/my-cat/checker.sh [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+if [ $POINTS -eq 90 ]; then
+    POINTS=$((POINTS + 10))
+    echo "Final Score: $POINTS/$TOTAL points"
+    echo "Good job!"
+else
+    echo "Final Score: $POINTS/$TOTAL points"
+fi
+echo "----------------------------------------"
+
+exit 0
```
Comment (teodutu) [inline chapters/io/file-descriptors/drills/tasks/my-cat/checker.sh:88]:
```suggestion
```

File: chapters/io/ipc/drills/tasks/anon-pipes/checker.sh [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+INPUT_FILE="input.txt"
+
+# Create input file if it doesn't exist
+if [ ! -f "$SOURCE_DIR/$INPUT_FILE" ]; then
+    touch "$SOURCE_DIR/$INPUT_FILE"
+fi
+
+# Read from terminal and write to input file
+read -r -p "Enter text to pass from parent to child: " input_text
+echo -e "\n$input_text\n" > "$SOURCE_DIR/$INPUT_FILE"
```
Comment (teodutu) [inline chapters/io/ipc/drills/tasks/anon-pipes/checker.sh:17]: Don't make the script interactive. Place the input string in the script. Use 2 strings: a short one for debugging and a longer one (more than 8K to fill the internal buffer) to test correctness.

File: chapters/io/file-descriptors/drills/tasks/my-cat/checker.sh [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        POINTS=$((POINTS + 30))
+        echo "PASSED (+30 points)"
+    else
+        echo "FAILED (+0 points)"
+    fi
+    rm test1.txt test2.txt test_output.txt test_out_2.txt
+}
+
+# Test 3: Empty file
+test_empty_file() {
```
Comment (teodutu) [inline chapters/io/file-descriptors/drills/tasks/my-cat/checker.sh:56]: This test passes even with the skeleton code. Change the solution and remove the code that `open()`s the input file. Turn it into a `TODO`.

File: chapters/io/file-descriptors/drills/tasks/mmap_cp/checker.sh [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    exit 1
+fi
+
+# Clean up
+rm testfile.txt testfile_mmap.txt testfile_rw.txt
+
+# Remove duplicate directory
+rm -rf ../../../../chapters/data/process-memory/drills/tasks/copy/
+
+exit 0
```
Comment (teodutu) [inline chapters/io/file-descriptors/drills/tasks/mmap_cp/checker.sh:27]:
```suggestion
```
Remove this from every checker script. It's OK for the scripts to fail

File: chapters/io/file-descriptors/drills/tasks/mmap_cp/checker.sh [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+./support/mmap_cp testfile.txt testfile_mmap.txt
+
+# Copy using standard cp command
+cp testfile.txt testfile_rw.txt
+
+# Verify the files are identical
+if cmp -s testfile.txt testfile_mmap.txt && cmp -s testfile.txt testfile_rw.txt; then
+    echo "File copies are identical."
+else
+    echo "File copies differ."
```
Comment (teodutu) [inline chapters/io/file-descriptors/drills/tasks/mmap_cp/checker.sh:17]: Add "test passed" and "test failed" to these messages to clarify when the output is correct and when it isn't.

**Review by teodutu — CHANGES_REQUESTED** [review]
Thank you for the scrips. Here's how to improve them.

- Move each checker to `solution/tests/` so that `make skels` places it in `support/tests`. Make sure to run `chmod +x support/tests/checker.sh` in the `skels` Makefile rule. And change the commands in the scripts to refer to their new paths.
- Move the solution code to `solution/src` so that `make skels` places its output in `support/src`.
- And make the original checkers executable.
- Add a CLI snippet where you run the checker script that passes all tests in each task's `README.md`.
- Change your commit message to `chapters/io: Add checker infrastructure for lab 9` followed by a description mentioning the tasks that received checkers.

**Comment by robertpaulp** [comment]
I've added your suggestions.

**Comment by teodutu** [comment]
Squash your 2nd commit into the first one. I'm not sure how you got it, but to merge the `main` branch you should do the following:

```
$ git checkout main
$ git fetch upstream  # first set your upstream remote to https://github.com/cs-pub-ro/operating-systems
$ git pull upstream main
$ git checkout -  # - is the previous branch
$ git rebase main  # Fix all conflicts, git add the conflicting files, then git rebase --continue and repeat until all commits are rebased
```
This way you won't get that extra commit that you now need to squash

**Comment by robertpaulp** [comment]
@teodutu Done.

**Review by Viktor09 — APPROVED** [review]

**Review by teodutu — CHANGES_REQUESTED** [review]
There's only one merge conflict that you added without solving. See my inline comment and then we can merge this. GJ!

**Review by teodutu — APPROVED** [review]
GG!

**Pull request merged**